Show line numbers in vty frontend

talanis85 opened the pull request:
Hello everyone!

This adds vim-style line numbers in vty frontend (as of #1054).

Currently it's not configurable (it displays line numbers for all non-mini windows), because I'm not exactly sure where to configure this. I think in vim, "number" is a property of the window but it can also be set globally, such that new windows are created with numbers on or off. So my guess would be to add fields to the Config and Window records and then create new windows with the number setting according to the number setting in Config.

Feedback appreciated :)

noughtmare commented:
Hello and thank you for your contribution!

I have not tested your changes yet, but it looks good.

The easiest way to make it configurable is to use a dynamic variable. Yi has a mechanism of `YiVariable`s where you can use the `getEditorDyn`, `putEditorDyn`, `getBufferDyn` and `putBufferDyn` functions to get and put dynamic variables.

I personally think it is nicer if every option is specified explicitly instead of using the dynamic state, so the global option would go in the [UIConfig record](https://github.com/yi-editor/yi/blob/b60347803c43f8502d57048a979269baff5da3ca/yi-core/src/Yi/Types.hs#L383) and the local options would go into the [Window record](https://github.com/yi-editor/yi/blob/b60347803c43f8502d57048a979269baff5da3ca/yi-core/src/Yi/Window.hs#L27). The disadvantage is that some function that use the `UIConfig` or `Window` records need to be updated as well. So you should probably not start off with this.

(A nice way to get the advantages of both the explicit and the dynamic methods is outlined in the functional pearl: http://www.cs.ru.nl/~W.Swierstra/Publications/DataTypesALaCarte.pdf and more recently: https://www.microsoft.com/en-us/research/wp-content/uploads/2016/11/trees-that-grow.pdf, but sadly none of that is used in Yi, because Yi is older than both those papers and it would be a lot of work to change it now)

When you have implemented the variables for the vim keymap you can look at how the paste option is implemented:  https://github.com/yi-editor/yi/blob/b60347803c43f8502d57048a979269baff5da3ca/yi-keymap-vim/src/Yi/Keymap/Vim/Ex/Commands/Paste.hs, it is a boolean option that is implemented in dynamic state.

I'm not sure what the best way is to make this configurable with the emacs keymap, but you could always just introduce a new keybinding that toggles the line numbers.

talanis85 commented:
Thank you for the detailed suggestions!

To keep things simple, I go for the YiVariable approach here. The downside is that you can only have numbers in all windows or no numbers at all. But vim has that general concept of buffer-local and global variables which Yi doesn't, so I think simple is better here.

I also added a vim ex command ":set number/nonumber" to toggle it.

noughtmare commented:
@talanis85 I think you can actually use `getBufferDyn` and `putBufferDyn` to get and set buffer local variables.

So I would add something like this to `Yi.UI.LineNumbers`:
```haskell
-- | A YiVariable which locally toggles line numbers for frontends
-- that support them.
-- Nothing means use global settings.
-- Just True means show line numbers in the current buffer. 
-- Just False means don't show line numbers in the current buffer.
newtype DisplayLineNumbersLocal = DisplayLineNumbersLocal { getDisplayLineNumbersLocal :: Maybe Bool }
  deriving (Generic, Typeable)

instance Default DisplayLineNumbersLocal where
  def = DisplayLineNumbersLocal Nothing

instance Binary DisplayLineNumbersLocal

instance YiVariable DisplayLineNumbersLocal
```

Then change `displayLineNumbers` in `Yi.Frontend.Vty` to:
```Haskell
displayLineNumbers = case getDisplayLineNumbersLocal $ snd $ runEditor cfg' getEditorDyn of
  Nothing -> getDisplayLineNumbers $ snd $ runEditor cfg' getEditorDyn e 
  Just b -> b
```

talanis85 commented:
Updated with a buffer-local version of `DisplayLineNumbers`. `putBufferDyn` is actually exactly what I wanted. Thanks for the pointer!

The semantics are still slightly incompatible with vim (which uses a window-local setting), but I feel that attaching the setting to a buffer instead of a window makes more sense anyway.

noughtmare commented:
Nice!

How easy would it be to also add options to the emacs keymap? Maybe some common functionality for setting and getting the variables can be moved to `Yi.UI.LineNumbers`, so that emacs can also use them in the future.

About the vim compatibility, we don't strive for full vim compatibility. Maybe you could add something to https://github.com/yi-editor/yi/blob/master/yi-keymap-vim/README.rst.

talanis85 commented:
As far as I know, emacs does line numbers using a minor mode. Since yi does not have that, I'm not sure what an emacs user would consider the "right way" to toggle such a setting.

Anyway, to make things easy I have changed `Yi.UI.LineNumbers` to only export getters and setters for both options. This has the added benefit that users of the module don't have to remember to use the correct `putXXXDyn` for each setting.

Now, getting and setting the option is just
```
b <- getDisplayLineNumbers[Local]
```
and
```
setDisplayLineNumbers[Local] b
```

To configure a default in your configuration, you have to add a startAction
```
EditorA (setDisplayLineNumbers [True|False])
```

noughtmare commented:
Great work!

I still have some suggestions:
* Make the line numbers a different color than the text
* Set a minimum width for the line numbers like in vim, i.e. make it (the `_` denotes a space):
```
__1_Some_text
__2_Some_text
__3_Some_text
__4_Some_text
``` 
instead of 
```
1_Some_text
2_Some_text
3_Some_text
4_Some_text
```
That way the width of the line numbers will only change after 1000 lines and not after 10 and 100.

* Make the global variable a field in the `Config` record, that way users won't have to use a `startAction` to enable line numbers by default. I don't know how difficult this will be, but I think it is worth a try. This wil also makes it possible to define proper lenses for the global line number settings like in the [Yi.Config.Simple](https://hackage.haskell.org/package/yi-core-0.17.1/docs/Yi-Config-Simple.html).

I think it is very nice that it is also possible to set line numbers per mode by doing:
```haskell
myMode { modeOnLoad = modeOnLoad myMode *> setDisplayLineNumbersLocal (Just True) }
```
That way you can for example only enable line numbers in haskell buffers.

That said, I think this pull request is ready for merging, so if you want I can merge it right now and then we will just open a new issue for each of the suggestions above.

talanis85 commented:
Yeah, the global setting should definitely be in `UIConfig`!

I've removed the global `YiVariable` and added `configLineNumbers` to `UIConfig`. I've also removed the somewhat strange `:set globalnumber` ex command.

Feel free to merge this now. I agree that it's a good idea to open a separate issue for the aesthetical details, since they are somewhat opinionated. Plus I first want to look into how themes work in yi.
